server/src/print.ts

```diff
@@ -40,6 +42,47 @@ export function typeToString(syn: ast.AnyType): string {
     }
 }
 
+export function parens(s: string): string {
+    return `(${s})`;
+}
+
+// To avoid confusion, we will assume 
+// 1. logical operators && and || have equal precedence
+// 2. bitwise (binary) operators have the equal precedence 
+// so we add extra parens in these cases even if unnecessary
+
+export function create_opmap() {
+    const opmap = new Map();
+    opmap.set("*", 1);
+    opmap.set("/", 1);
+    opmap.set("%", 1);
+    opmap.set("+", 2);
+    opmap.set("-", 2);
+    opmap.set("<<", 3);
+    opmap.set(">>", 3);
+    opmap.set("<", 4);
+    opmap.set(">", 4);
+    opmap.set("<=", 4);
+    opmap.set(">=", 4);
+    opmap.set("==", 5);
+    opmap.set("!=", 5);
+    opmap.set("&", 6);
+    opmap.set("^", 7);
+    opmap.set("|", 8);
+    opmap.set("&&", 9);
+    opmap.set("||", 9);
+
+    return opmap;
+
+}
+
+export function cmp_precedence(o1: string, o2: string) {
+    const opmap = create_opmap();
+    if (opmap.get(o1) > opmap.get(o2)) return -1;
```

Comment:
Can you move the ordering enum from `ast-search` into `util` and use that instead?